drivers/sensor/adi/adxl34x/adxl34x_decoder.c

```diff
@@ -0,0 +1,302 @@
+/*
+ * Copyright (c) 2024 Chaim Zax
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/sensor/adxl34x.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+#include "adxl34x_decoder.h"
+#include "adxl34x_convert.h"
+
+LOG_MODULE_DECLARE(adxl34x, CONFIG_SENSOR_LOG_LEVEL);
+
+#if defined(CONFIG_ADXL34X_DATA_TYPE_Q31)
+
+/**
+ * Convert the raw sensor data from the registery to the q31 format
+ *
+ * @param[in] value The raw value to convert
+ * @param[in] range_scale The scale of range
+ * @param[in] shift The shift value to use in the q31 number
+ * @param[out] out Pointer to the converted q31 number
+ */
+static void adxl34x_convert_raw_to_q31(const uint8_t *value, uint16_t range_scale, int8_t shift,
+				       q31_t *out)
+{
+	const int16_t raw_value = sys_get_le16(value);
+	const int32_t ug = (int32_t)raw_value * range_scale * 100;
+	struct sensor_value ms2;
+
+	sensor_ug_to_ms2(ug, &ms2);
+	const int64_t q31_temp = ((int64_t)ms2.val1 * INT64_C(1000000) + ms2.val2) *
+				 ((int64_t)INT32_MAX + 1) / ((1 << shift) * INT64_C(1000000));
+	*out = CLAMP(q31_temp, INT32_MIN, INT32_MAX);
```

Comment:
This should be a simple mulitply rather than requiring division, shifting, and 64 bit math. The raw value is 16bit, you are converting to a q1.31 with a new scale.

See #76968

Reply:
Thanks for the feedback @teburd, I was starting to worry if anyone would take the time for this review. 

I am new with the q31 notation so I mainly looked at existing drivers for examples. I see what you're getting at. No problem, the driver already uses some lookup tables for conversion and it shouldn't be a problem to extend this in favor of having these calculations done runtime.